cmd/stellar-rpc/internal/ingest/backfill.go

```diff
@@ -0,0 +1,278 @@
+	if !dbIsEmpty && (maxWrittenLedger >= minWrittenLedger) {
+		logger.Infof("Starting precheck for backfilling the database at %s, phase 1 of 4", localDbPath)
+		if err = verifyDbGapless(ctx, localDbReader, minWrittenLedger, maxWrittenLedger); err != nil {
+			return fmt.Errorf("backfill precheck failed: %w", err)
+		}
+	} else {
+		logger.Infof("Local DB is empty, skipping precheck")
+	}
+	logger.Infof("Precheck passed! Starting backfill backwards, phase 2 of 4")
+
+	// Phase 2: backfill backwards towards oldest ledger to put in DB
+	if err := getLatestSeqInCDP(ctx, dsInfo.Ds, &currentTipLedger); err != nil {
+		return fmt.Errorf("could not get latest ledger number from cloud datastore: %w", err)
+	}
+	logger.Debugf("Current tip ledger in cloud datastore is %d", currentTipLedger)
+	lBound := max(currentTipLedger-nBackfill+1, 1)
+
+	var rBound uint32
+	if dbIsEmpty {
+		rBound = currentTipLedger
+	} else {
+		rBound = minWrittenLedger - 1
+	}
+	if err = runBackfillBackwards(ctx, logger, localDbRW, dsInfo, lBound, rBound); err != nil {
+		return fmt.Errorf("backfill backwards failed: %w", err)
+	}
+
+	// Phase 3: backfill forwards towards latest ledger to put in DB
+	logger.Infof("Backward backfill of old ledgers complete! Starting forward backfill to current tip, phase 3 of 4")
+	if dbIsEmpty {
+		lBound = rBound + 1
+	} else {
+		// If the DB wasn't empty initially, the backwards backfill filled up to minWrittenLedger-1 < maxWrittenLedger
+		lBound = maxWrittenLedger + 1
+	}
+	if err = getLatestSeqInCDP(ctx, dsInfo.Ds, &currentTipLedger); err != nil {
+		return fmt.Errorf("could not get latest ledger number from cloud datastore: %w", err)
+	}
+	if err = runBackfillForwards(ctx, logger, localDbRW, dsInfo, lBound, currentTipLedger); err != nil {
+		return fmt.Errorf("backfill forwards failed: %w", err)
+	}
+
+	// Phase 4: verify no gaps in local DB after backfill
+	logger.Infof("Forward backfill complete, starting post-backfill verification")
+	// Note final ledger we've backfilled to
+	endSeq := currentTipLedger
+	if err = getLatestSeqInCDP(ctx, dsInfo.Ds, &currentTipLedger); err != nil {
+		return fmt.Errorf("could not get latest ledger number from cloud datastore: %w", err)
+	}
+	startSeq := max(currentTipLedger-nBackfill+1, 1)
+	if err = verifyDbGapless(ctx, localDbReader, startSeq, endSeq); err != nil {
+		return fmt.Errorf("post-backfill verification failed: %w", err)
+	}
+	logger.Infof("Backfill process complete")
+
+	return nil
+}
+
+// Checks to ensure state of local DB is acceptable for backfilling
+func verifyDbGapless(callerCtx context.Context, reader db.LedgerReader, minLedgerSeq uint32, maxLedgerSeq uint32) error {
+	ctx, cancelPrecheck := context.WithTimeout(callerCtx, 4*time.Minute)
+	defer cancelPrecheck()
+
+	tx, err := reader.NewTx(ctx)
+	if err != nil {
+		return fmt.Errorf("db verify: failed to begin read transaction: %w", err)
+	}
+	defer tx.Done()
+
+	ct, err := tx.CountLedgersInRange(ctx, minLedgerSeq, maxLedgerSeq)
```

Comment:
really no reason to shorten it here
```suggestion
	count, err := tx.CountLedgersInRange(ctx, minLedgerSeq, maxLedgerSeq)
```